design/ceph/cephx-keyring-rotation.md

```diff
@@ -0,0 +1,438 @@
+    5. For each mgr, rotate the key and update/restart the mgr daemon
+    6. <!-- TBD --> Update the CephCluster status with the mgr key info (with internal retry)
+    7. For each OSD, rotate the key
+        1. OSDs don't have an associated keyring secret, so we must make sure all OSD deployments'
+            "activate" init container gets the latest key and updates it for the "osd" pod
+    8. <!-- TBD --> Update the CephCluster status with the OSD key info (with internal retry)
+
+With separated daemon key info tracking, the status will look like so:
+
+```yaml
+status:
+  # ...
+  cephx:
+    admin:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    mon:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    mgr:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    osd:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    csi: {} # discussed more below, but some CSI keys and need to be in this status
+```
+
+<!-- TODO: DISCUSS -->
+
+I believe we can track all CephCluster key rotations as one and rely on operator logs and error
+events for debugging, but we should still separate out CSI rotations since those can be initiated
+separately.
+
+```yaml
+status:
+  # ...
+  cephx:
+    local:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    csi: {} # discussed more below, but some CSI keys and need to be in this status
+```
+
+### CSI example
+
+This is complicated, and I need some input. My understanding is that CSI keys are are always updated
+as part of a CephCluster, CephRBDMirror, or CephFilesystemMirror reconcile.
+<!-- TODO: is the above true?
+    are any keys created in the CSI reconcile logic that's separate from CephCluster or other reconcile routines? -->
+
+For a CephCluster reconcile:
+
+1. Ceph-CSI daemons use a separate config from Ceph local, so the reconcile reads
+    `CephCluster.spec.security.cephx.csi` to determine if the reconcile should rotate.
+2. Assuming the rotation is indicated, rotate keys.
+3. Update the CephCluster status with the CSI key info (with internal retry)
+
+    ```yaml
+    kind: CephCluster
+    # ...
+    status:
+      # ...
+      cephx:
+        # (status from CephCluster local rotations)
+        csi:
+          keyGeneration: 2 # e.g.
+          keyCephVersion: "20.2.1" # e.g.
+    ```
+
+`(with internal retry)` this is an important detail we should plan for IFF the CSI reconciliation
+routine may be trying to update this value in parallel with CephCluster reconciliation. If there is
+no parallelism, the internal retry can (should) be removed from the design.
+
+If CSI also requires CephRBDMirror and CephFilesystemMirror reconciles to rotate keys, Rook must
+ensure those reconcile watchers enqueue for `CephCluster.spec.security.cephx.csi` updates. Other
+reconcile watchers can ignore `...cephx.csi` updates.
```

Comment:
IMO this is not required we can skip it. CSI and Mirroring is not tied to each other.